src/Initializer/ParameterDB.cpp

```diff
@@ -250,13 +331,131 @@ namespace seissol {
+      if (const ElementAverageGenerator* gen = dynamic_cast<const ElementAverageGenerator*>(&queryGen)) {
+        const unsigned numElems = numPoints / NUM_QUADPOINTS;
+        std::array<double, NUM_QUADPOINTS> quadratureWeights{ gen->getQuadratureWeights() };
+        std::vector<double> elemVolumes{ gen->getElemVolumes() };
+
+        // Divide by volumes to obtain parameter mean values and store them in m_materials
+        for (unsigned elementIdx = 0; elementIdx < numElems; ++elementIdx) {
+          m_materials->at(elementIdx) = this->computeAveragedMaterial(elementIdx, elemVolumes[elementIdx], quadratureWeights, materialsFromQuery); 
+        }
+      } else {
+      // Usual behavior without homogenization
+        for (unsigned i = 0; i < numPoints; ++i) {
+          m_materials->at(i) = T(materialsFromQuery[i]);
+        }
+      }
       delete model;
     }
+
+    // Computes the averaged material, assuming that materialsFromQuery, stores 
+    // NUM_QUADPOINTS material samples per mesh element.
+    // We assume that materialsFromQuery[i * NUM_QUADPOINTS, ..., (i+1)*NUM_QUADPOINTS-1] 
+    // stores samples from element i.
+    // Fallback variant: if no specialization is implemented return materialsFromQuery[i*NUM_QUADPOINTS]
+    // Note: this base variant should actually never be called, but only the specializations!
+    // The specializations should implement proper averaging.
+    template<class T>
+    T MaterialParameterDB<T>::computeAveragedMaterial(unsigned elementIdx, double elementVolume, std::array<double, NUM_QUADPOINTS> const& quadratureWeights, std::vector<T> const& materialsFromQuery) {
+      logInfo() << "You want me to compute an average material for a generic type. In general, this function should never be called, but always a proper specialization!";
+      unsigned globalPointIdx = NUM_QUADPOINTS * elementIdx;
+      return T(materialsFromQuery[globalPointIdx]);
+    }
+
+    template<>
+    seissol::model::ElasticMaterial MaterialParameterDB<seissol::model::ElasticMaterial>::computeAveragedMaterial(unsigned elementIdx, double elementVolume, std::array<double, NUM_QUADPOINTS> const& quadratureWeights, std::vector<seissol::model::ElasticMaterial> const& materialsFromQuery) {
+      double muMeanInv = 0.0;
+      double rhoMean = 0.0;
+      double vERatioMean = 0.0;
+      for (unsigned quadPointIdx = 0; quadPointIdx < NUM_QUADPOINTS; ++quadPointIdx) {
+        double quadWeight = 6 * elementVolume * quadratureWeights[quadPointIdx];
+        unsigned globalPointIdx = NUM_QUADPOINTS * elementIdx + quadPointIdx;
+        auto& localMaterial = materialsFromQuery[globalPointIdx];
+        muMeanInv += 1 / localMaterial.mu * quadWeight;
+        rhoMean += localMaterial.rho * quadWeight;
+        vERatioMean += localMaterial.lambda / (2 * localMaterial.mu * (3 * localMaterial.lambda + 2 * localMaterial.mu)) * quadWeight;
+      }
+
+      rhoMean /= elementVolume;
```

Comment:
it seems you multiply by elementVolume in quadWeightand divide here. Is it really necessary?

Reply:
You are right, I entirely removed the element volume calculations